### PR TITLE
fix: restructure use of rayon to avoid deadlocks

### PR DESCRIPTION
**Pull request by dignifiedquire**
- perf: Uses rayon in the places where the work is cpu bound only 
- fix: Do not use custom rayon thread pools, as multiple of these can result in cross pool deadlocking.
- fix: Proxy `BELLMAN_NUM_CPUS` to `RAYON_NUM_THREADS` for now, and deprecate it.
- refactor: Use `RAYON_NUM_THREADS` as the baseline for determining the number of threads to to use
- perf: Use `in_place_scope` instead of `scope` to avoid spawning unnecessary threads.
- ci: Add CI tests to ensure the currently deadlocking test for `groth16_bench` is checked

For reference
- https://github.com/rayon-rs/rayon/issues/751
- https://github.com/wasmerio/wasmer/issues/2261#issuecomment-826944128

### CODE REVIEW

**Review by vmx — COMMENTED** [review]
Code wise it looks good to me, but CI sadly isn't happy.

**Comment by dignifiedquire** [comment]
> Code wise it looks good to me, but CI sadly isn't happy.

Hmm, it passes locally, so I have a suspicion the test case is too large for CI

File: .circleci/config.yml [hunk 1]
```diff
@@ -92,6 +92,21 @@ commands:
...
 
+  test_target_blst_gpu_ignored:
+    parameters:
+      target:
+        type: string
+    steps:
+      - *restore-workspace
+      - *restore-cache
+      - run:
+          name: Test blst ignored (GPU) (<< parameters.target >>)
```
Comment (cryptonemo) [inline .circleci/config.yml:103]: Should probably take GPU out of the test name (and also above)

File: tests/groth16_aggregation.rs [hunk 2]
```diff
@@ -306,9 +306,9 @@ fn test_groth16_bench() {
     let inclusion = vec![1, 2, 3];
     for i in nb_proofs {
         let mut records = Vec::new();
+        let (pk, vk) = generic.specialize(i);
         for _ in 0..n_average {
             println!("Proofs {}", i);
-            let (pk, vk) = generic.specialize(i);
```
Comment (cryptonemo) [inline tests/groth16_aggregation.rs:311]: Ouch. Good catch.

File: .circleci/config.yml (hunk 1 above)
Comment (dignifiedquire) [reply .circleci/config.yml]: done

**Pull request merged**